### DISCUSSION TITLE
RS232 on a Commodore 64 using DurexForth

### DISCUSSION THREAD

**Discussion by AussieSmitty** [jkotlinski/durexforth#330] [General]
Hi all, I have managed to use the RS232 adapter (VIC-1011A) to talk to a PC and Arduino (using RS232 shield), using both BASIC and Assembler:
10 PRINT CHR$(147);
20 OPEN 5,2,1,CHR$(8)
30 GET#5,A$
40 IF A$<>"" THEN PRINT A$;
50 GET B$
60 IF B$<>"" THEN PRINT#5,B$;
70 GOTO 30
RSINI:   LDA #$02
         LDX #<PAR
         LDY #>PAR
         JSR SETNAM
         LDA #$02
         TAX
         LDY #$00
         JSR SETLFS
         JSR OPEN
         ; TRANSMISSION
         LDX #$02
         JSR CHKOUT
         LDA X_Var   ; Where the byte to send is in X_Var
         JSR CHROUT
PAR:     .BYTE %00001000,%00000000

Here is my code in DurexForth, but no bytes are being sent:
( RS232 on C64 testing )
: setcom ( -- ) [
$fb stx,   \ store a copy of xreg
8 lda,#    \ low byte of params
$fd sta,   \ store at $fd
0 lda,#    \ hi byte of params
$fe sta,   \ store at $fe
2 lda,#    \ length of filename
$fd ldx,#  \ lo byte of params add
$fe ldy,#  \ hi byte of params add
$ffbd jsr, \ call kernal setnam
2 lda,#    \ 2 = file# & RS232 device#
tax,
0 ldy,#    \ no secondary #
$ffba jsr, \ setlfs
$ffc0 jsr, \ open
$fb ldx, ] ; \ restore xreg

: sendbyte ( $fc -- ) [
$fb stx,    \ store current val in xreg
2 ldx,#     \ rs232 device #
$ffc9 jsr,  \ chkout
$fc lda, \ byte to send
$ffd2 jsr,  \ chrout
3 ldx,#     \ screen device #
$ffc9 jsr,  \ chkout
$fb ldx, ] ; \ restore xreg

: main ( -- ) setcom
147 emit 62 emit 13 emit
10 0 do key dup
$fc c! sendbyte emit loop ;
Any suggestions where I could be going wrong? Thanks.

**Comment by burnsauce** [comment]
Edit: just saw the brackets. Nevermind.

**Reply by AussieSmitty** [reply]
Hi Burnsauce, Thanks for at least having a look. It wouldn't be the first time my code has an obvious bracket missing.

**Comment by jkotlinski** [comment]
Maybe needs a CLRCHN following CHROUT?

**Reply by AussieSmitty** [reply]
Thanks, jkotlinski, that's worth checking out. I'll let you know if that works.

**Reply by AussieSmitty** [reply]
Hi Jkotlinski, while adding CLRCHN might be good housekeeping in a Commodore 64, to get RS232 working with DurexForth doesn't need this added. As you've seen from Whammo, DurexForth modifies the NMI vector and therefore all RS232 handling is being bypassed. Hopefully this can be corrected in later releases of DurexForth, but for now Whammo has provided a work-around that seems to be working for me for now.

**Comment by Whammo** [comment]
I helped someone with this years ago and never heard back from them, so I assumed it worked.
The NMI vector is changed as DurexForth is initialized, so perhaps if it is reset to the default before the OPEN, we might have better luck.

**Reply by AussieSmitty** [reply]
Thanks, Whammo, I also wondered whether DurexForth modified the memory map, but hadn't considered it may also affect the MMI vector. I'll let you know if I can figure this out with some C@ . of suitable memory locations. It makes sense this might be why I'm not seeing any bytes being transmitted with RS232.

**Reply by Whammo** [reply]
Interpreter.asm shows that the NMI vector runs through QUIT. QUIT resets the NMI vector to the restore handler that calls QUIT.
 Pretty heady stuff.

**Reply by AussieSmitty** [reply]
That sounds a bit wierd, and I'm struggling to think of how I can solve this issue. Perhaps I could load a 3rd party RS232 handler (some exist since the original code Commodore put into the C64 couldn't reliably go to 9600 baud for instance), and try to figure out a way to have this called as needed for transmission and reciept over RS232.

**Comment by jkotlinski** [comment]
Wait. Does NMI cause problems for RS-232?
It never occured to me that it was useful for anything but handling run
stop/restore.

**Reply by Whammo** [reply]
RS-232 bit timing is driven by NMI CIA timer interrupts. At this moment I don't know what the KERNAL does with the restore handler vector or the default when RS-232 is OPENed but if it falls through at some point to QUIT, it's not gonna be good for RS-232.

**Reply by Whammo** [reply]
![image](https://user-images.githubusercontent.com/38296423/110556740-fcb41280-80f3-11eb-9b56-cb8554c8f845.png)

**Reply by AussieSmitty** [reply]
Thanks Whammo, this looks helpful (I think).

**Reply by Whammo** [reply]
At this point I am fairly certain that RS-232 open doesn't change the NMI vector. DurexForth certainly does.  
 The NMI vector defaults to $FE47 so to reset the NMI:   

$FE47 $318 !

This will cause the restore key to have 'undefined' behavior, but RS-232 may work.

**Reply by AussieSmitty** [reply]
Thanks again Whammo, that certainly seems logical, so I'll give this a try tonight (not with my C64 setup atm). I will also report back my results.

**Reply by AussieSmitty** [reply]
Hi Whammo, I tried the $FF47 $318 !  and it worked! I haven't had a chance to digest your further suggestion, but for now I'm really pleased I can talk via DurexForth to an external RS232 device. 
I will now see about receiving bytes as well, and then have a look at your suggestion around keeping the NMI handling DurexForth wants to do, but still support RS232.
Thanks again so much for this as I was quite frustrated with not being able to move forward with my intended robotic projects (they will all benefit from 2 way comms, and RS232 is a good way to achieve this with a Commodore 64 and external 'modern' devices)
Regards,
Steve

**Reply by Whammo** [reply]
That's great to hear! I appreciate the work you've done to test it out.

**Comment by jkotlinski** [comment]
It definitely sounds like DurexForth NMI handler should check if Run/Stop
is pressed before calling QUIT.

**Reply by Whammo** [reply]
I've been looking at the code and I think that if we ~instead redirect the BRK vector to QUIT~, and leave the NMI vector at $FE47, 
this would then catch stop/restore to quit, and restore would likely init a cart unless interrupts are enabled and then it bypasses the cart, and stop key check.

EDIT- Sorry no, because stop key check doesn't go through the BRK vector it just falls through.

**Reply by AussieSmitty** [reply]
Hi Whammo,
Sorry, so is this code something I should try or your simple trick of $FE47 $318 ! is all I need? Is being able to use the Stop key helpful? In my assembler programming I've never used the Stop key, but is it a different story with DurexForth?
Ultimately, I hope to be able to write code that could then be run by a user who wouldn't know Forth was under the covers. Can DurexForth be "compiled" into some sort of stand-alone runable code?
Sorry, lot's of questions but I really like Forth.
Thanks again for being so helpful.

**Reply by Whammo** [reply]
I think that if it works okay the way it is, that's great.

**Comment by Whammo** [comment]
This might be fun. Set this to NMI vector. Set QUIT to the BRK vector.
  It skips starting the cart on restore, and JMPs through the BRK vector on run/stop restore.
 It *may* mess with timing but I doubt it.
  


```forth
code nmi
        pha,	\ save a
        txa,	\ copy x
        pha,	\ save x
        tya,	\ copy y
        pha,	\ save y
    $7f lda,#   \ disable all interrupts
  $dd0d sta,  	\ save via 2 icr
  $dd0d ldy,    \ save via 2 icr
+branch bpl,    \ skip if nmi enabled
  $fe72 jmp,    \ go directly to rs-232 routines
:+ 
  $f6bc jsr,	\ increment real time clock
  $ffe1 jsr,	\ scan stop key
+branch	bne,    
  $fe72 jmp,	\ if not [stop] restore registers and exit interrupt
:+
        pla,
        pla,
        tax, \ still need to restore X
  $316  (jmp), \ brk vector
;code

: install ' quit $316 !  ' nmi $318 ! ;
```
NOTE; Actually, QUIT does reset the return stack.

**Reply by AussieSmitty** [reply]
Hi Whammo,
Thanks for this. I sent a reply to your earlier suggestion and this just arrived. Perhaps I will try this only once I appreciate why I need to support QUIT (which I assume is invoked by pressing the Stop key?).
Regards,
Steve

**Reply by Whammo** [reply]
I guess QUIT is important in that you can write code that blocks, you know, waits for something to happen. Then you can break in if it doesn't happen without resetting the machine. I'm sure it's more than that too.

**Reply by jkotlinski** [reply]
No, I think that's pretty much it :)  If there is an endless loop, or there is some crash or other weirdness due to mistake in the Forth code, Run Stop+Restore should reset the Forth to get the interpreter back in a working state.

**Comment by jkotlinski** [comment]
It is nice to have the QUIT on runstop+restore, because then its possible
to exit endless loops.
But for sure this feature should not prevent RS-232 usage.
That should be reported as an issue and fixed.

**Reply by AussieSmitty** [reply]
Hi Jkotlinski,
How does one report an issue like this to get fixed?
Regards,
Steve

**Reply by jkotlinski** [reply]
Hallo! Whammo already created an issue + potential fix :-)
https://github.com/jkotlinski/durexforth/issues/331

**Comment by Whammo** [comment]
Here is a patch for RS-232 it ignores restore, and carts, quits on run/stop restore and even handles BRK nicely.

nmi.fs

```forth
$ce1 value rsh
$cdc value rsl
' quit value out
code nmi
        pha,    \ save a
        txa,    \ copy x
        pha,    \ save x
        tya,    \ copy y
        pha,    \ save y
    $7f lda,#   \ disable all interrupt
  $dd0d sta,    \ save via 2 icr
  $dd0d ldy,    \ save via 2 icr
+branch bpl,    \ skip if nmi enabled
  $fe72 jmp,    \ go directly to rs-232
:+
  $f6bc jsr,    \ increment real time c
  $ffe1 jsr,    \ scan stop key
+branch bne,
  $fe72 jmp,    \ if not [stop] restore
:+
        pla,
        pla,
        tax,
out   jmp,
;code

code brk
         pla,
         pla,
         tax,
out   jmp,
;code

' brk $316 !  ' nmi $318 !
$316 c@ rsl c! $317 c@ rsh c!
quit
```

**Reply by AussieSmitty** [reply]
Hi Whammo,
Good idea to check the RS232 buffer 
I'll try that tomorrow night

**Reply by AussieSmitty** [reply]
Hi Whammo,
My code seems to be working quite well. The 10 characters that are supposed to be transmitted are indeed in the RS232 output buffer, but the index to the end of output buffer is reading 10, meaning the kernal routine that is supposed to be sending these 10 characters must not have sent any. Strangely I do get some intermittent (yet consistent) bytes arriving over the RS232 at the destination but not the characters that are in the RS232 output buffer.
![image](https://user-images.githubusercontent.com/41884798/113128880-ff72d500-9265-11eb-85e2-b127537770ec.png)

Just to make sure there was no issues with my setup, I then did a hardware reset (in WinVICE) and loaded the basic program I used previously and confirmed all characters typed were being sent via RS232 successfully.

![image](https://user-images.githubusercontent.com/41884798/113130293-a4da7880-9267-11eb-98c3-299f3b13a89d.png)
The middle 4 bytes is me typing TEST in my basic program. The 254 254 255 255 is all I get out of DurexForth after typing 10 characters "test123456" in my Forth code.

**Comment by Whammo** [comment]
Is the NMI vector being changed back from $FE47?

**Reply by AussieSmitty** [reply]
Hi Whammo,
No. After I performed the $FE47 $318 !  I assumed that's all that I needed to do. Are you saying that after my code I need to change something back again?
Thanks for your assistance.
Steve

**Reply by Whammo** [reply]
I'm trying to determine whether or not durexforth is changing the NMI vector back to point at QUIT at some point, because it will do that under certain conditions. So what I am saying after it fails to send, do $318 @  and see if it's still $FE47

**Comment by Whammo** [comment]
Fundamentally, the issue with RS-232 on DurexForth is the handling of NMI.
 ```assembly
restore_handler
    cli
    jmp QUIT

quit_reset
    sei
    lda #<restore_handler
    sta $318
    lda #>restore_handler
    sta $319
    cli
```
 These are the first few lines of interpreter.asm. quit_reset is executed upon startup, through QUIT and through restore_handler.
As you can see, quit_reset sets the NMI vector to restore_handler. 
```forth
$ce1 value rsh \ from my patch above
$cdc value rsl
```
These are the addresses in the current distro <restore_handler and >restore_handler following their respective lda in quit_reset.
 If these addresses are changed to $FE47, 
```forth
$FE rsh c! $47 rsl c!  
``` 
 it might eliminate this as an issue.

**Reply by AussieSmitty** [reply]
Hi Whammo,
Thank you again for your dedication to helping solve this issue. What you've suggested is certainly worth trying, so as soon as I'm able this evening (my time) I'll give this a try and let you know how this goes.
Regards,
Steve

**Reply by Whammo** [reply]
I'm curious if you've used the patch I've posted.

**Reply by AussieSmitty** [reply]
No, I assumed the $FE47 $318 ! idea was all I had to do. The 'patch' you posted, do I run this inside my code or once only to try and fix the issue?

**Reply by Whammo** [reply]
Ah. Yes, just include it once before anything else.

**Reply by AussieSmitty** [reply]
Sorry, I hadn't tried this yet. I will do and report my outcome. Looks like it should work (I think).

**Reply by AussieSmitty** [reply]
Hi Whammo,
The good news is I'm now not getting the weird 254 and 255 bytes sneaking out via RS232 from DurexForth in my testing, however I'm now getting nothing instead. I tried with and without doing the $FE37 $318 ! after having 'run' your patch (hitting F7 in vi after pasting in the code).
![basic test rs232 to arduino](https://user-images.githubusercontent.com/41884798/113255410-89c54280-9313-11eb-9e54-332b6536e49b.JPG)
![DurexForth test rs232 to arduino](https://user-images.githubusercontent.com/41884798/113255464-9a75b880-9313-11eb-869a-620b4397e72d.JPG)

I'm also going to do some testing shortly using a real C64 just in case there is something wierd going in in WinVICE (I'm running v3.5, which is the latest, but you never know with emulators). But since I'm able to use BASIC and that works fine, I'm fairly convinced it's just DurexForth that's the issue.

Thanks again for any help you can offer.

**Reply by Whammo** [reply]
I'm glad to help. It's unfortunate, but as it is, DurexForth doesn't support rs232. I've dabbled with attempting to get it going with VICE, and had similar success. You may have better luck on an actual 64.
 I can't spare the bench time right now to attempt to resolve this once and for all, but I've made a good start of it, for someone who does, I think.
 You definitely do not want to  $FE37 $318 ! after having run the patch. <------- edit

**Comment by Whammo** [comment]
You also might want to try it with BASIC ROM flipped in.

**Reply by AussieSmitty** [reply]
Hi Whammo,

Using my real C64 produced the same results of no characters getting out from the C64 over RS232.

As you may recall, I also wrote and tested successfully some machine code that works with RS232, so I'm not sure switching BASIC ROM in will be a factor. I actually don't know how to do this in any case.

It's a shame I can't get this to work as I had great plans to do a lot of robots projects using DurexForth, and RS232 communications would be an ideal way to have 2-way data transfers, given how so many devices support this comm type.

Perhaps the way I am loading/compiling/invoking your patch isn't correct.

I have some forth words (as per my original message) followed by your code. When I hit F7 I assume the code not part of a word definition is automatically executed. I then 'run' my words, first SETCOM to set the RS232 registers for 1200 baud, 8bits, no parity. Then is use the word MAIN to send 10 characters as I type them and check if they are getting over to the Arduino (using it's monitor). So BASIC and Assembler codes work, but I can't get DurexForth to achieve the same result.

Let me know if I'm applying the patch how it should be loaded.

Regards,
Steve

**Reply by Whammo** [reply]
The thing is, I put considerable time and effort studying the solution and writing a patch based on the idea that $FE37 $318 ! resolved the problem. Since it doesn't, it's going to require considerably more. 
 At this time, all I know is that DurexForth changes the NMI vector and that's why rs232 doesn't work.

**Reply by AussieSmitty** [reply]
Hi Whammo,

Thank you again for all your effort and patience to this point.

I'm away for Easter but on my return I will make a concerted effort to better understand how I can work with the NMI code redection to serve both RS232 and what-ever DurexForth wants to do. I will report any progress I get.

Thanks again for the effort so far!

Steve

**Reply by AussieSmitty** [reply]
Early testing shows that the vector keeps getting reset after changing it, so that may be the issue. 
Still doing more testing.

**Reply by Whammo** [reply]
This will set and prevent the vector from changing:
```forth
$FE47 $318 ! $FE $ce1 c! $47 $cdc c!  
```

**Reply by AussieSmitty** [reply]
Hi Whammo,

Thanks, I will try this tonight and let you know how I go (both in WinVICE and on real hardware).

Steve

**Reply by AussieSmitty** [reply]
Hi Whammo,
Some success. I'm seeing a byte being transmitted for each key I type, but so far I'm only getting 255 or 254. I need to tidy what gets built (I still have your other rountine you provided previously, so it may be clashing with what we're trying to do.
I'm tired after a big day at work, but will get into this more tomorrow and report.
Steve

**Reply by AussieSmitty** [reply]
Hi Whammo.
I have only managed to get a short time this week to do more testing, and unfortunately I'm still not getting it to work. I tested on a real C64 as I wanted to confirm that WinVICE isn't the issue. I went back and made sure my BASIC and Assembler code programs work and they do, so it's getting DurexForth to work with RS232 that I'm still stuck with.
I will try to do some more testing this weekend as I want to make 100% sure that my Forth code isn't perhaps the issue (that could be embarrassing!). Currently I'm not using the stack to store the byte I want to send, which looks sloppy in Forth given that ideally the value to be sent should be popped on the stack first. Having said that, I can't see (yet) if there is something in my code that is preventing the actual bytes I want sent ending up as $ff or $fe.
Regards,
Steve

**Comment by Whammo** [comment]
If you've got it working in pure assembly outside of durexforth, there's no reason you cannot LOADB your binary, 
 use $FE47 $318 ! $FE $ce1 c! $47 $cdc c!  to fix NMI and SYS to your routines.

**Reply by AussieSmitty** [reply]
Hi Whammo,

OK, thanks I'll try this and let you know how I go.

Regards,
Steve

**Comment by jkotlinski** [comment]
Hi Steve,
If you have a possibility to share your project, maybe I could take a look
at it too, if I find the time :-)

**Reply by AussieSmitty** [reply]
Hi Jkotlinski,

That would be great. I've attached the DurexForth .D64 file containing the code I'm using to try to get RS232 working.

I'm only a hobby programmer, so perhaps it's my code that's the issue, but Basic and Assembler are working, so I at least know RS232 is working on either real hardware or via WinVICE.

Thanks for any help you can offer.

Regards,
Steve

**Reply by AussieSmitty** [reply]
Hi Jkotlinski,
I've just realised that this discussion trial can't accept .D64 files. What's the best way to get you this?
Regards,
Steve

**Reply by AussieSmitty** [reply]
Hi Jkotlinski,

How do I get you a copy of my .D64 DurexForth file?

Regards,
Steve

**Reply by Whammo** [reply]
It will accept a .zip file. just rename it to xxx.zip and point out its a  ,d64 and to rename it so.

**Reply by AussieSmitty** [reply]
Hi Jkotlinski,

I've turned the .D64 file into a ZIP file. 
[RobotGuitarForth.zip](https://github.com/jkotlinski/durexforth/files/6340528/RobotGuitarForth.zip)

Let me know if you got this OK?

Regards,
Steve

**Reply by AussieSmitty** [reply]
Hi Whammo,

I actually ZIPped the file up (using an online free ZIP utility). If needed, I will also try just renaming the extension.

Thanks,
Steve

**Reply by Whammo** [reply]
I've included a patch to keep the BASIC ROM in context. Rename file to rgf.d64


[rgf.zip](https://github.com/jkotlinski/durexforth/files/6341440/rgf.zip)

**Comment by Whammo** [comment]
I've changed some code, my testing indicates it may be working now.
(change filename to rgf0.d64)
[rgf0.zip](https://github.com/jkotlinski/durexforth/files/6342492/rgf0.zip)

**Reply by AussieSmitty** [reply]
Hi Whammo,

I'm just downloading and trying it now too. 

Regards,
Steve

**Comment by Whammo** [comment]
$9e00 is the output buffer. It shows two runs of MAIN.
$29d is the buffer head and $29e is the tail.
 They're the same so it's been transmitted.
I have also confirmed that the NMI register is being hit.



![rgf](https://user-images.githubusercontent.com/38296423/115383612-e84f4500-a18a-11eb-875e-3d2cf3aa9450.jpg)

**Reply by AussieSmitty** [reply]
Hi Whammo,

I just completed a test via WinVICE out to an Arduino and it's working!!! Awesome.

I'm about to also do a test on a real C64, but given this test's success, I can already predict it will work.

This is great stuff and well done, thank you.

Regards,
Steve
![Test of RS232-2](https://user-images.githubusercontent.com/41884798/115386557-e7ff8a00-a21c-11eb-98cc-ad1f60faa505.JPG)

**Reply by AussieSmitty** [reply]
Test also confirmed on real C64 now too!

**Comment by Whammo** [comment]
I am not certain that the BASIC ROM is required to be in, but I am curious,

**Reply by AussieSmitty** [reply]
Hi Whammo,

I just tried 

(Pound) basic out
 
And it works fine too!

Regards,
Steve

**Reply by Whammo** [reply]
Excellent! Thank you.

**Reply by AussieSmitty** [reply]
Hey Whammo, is there a way I can pay you for this service?

**Reply by Whammo** [reply]
Also, when you have a moment, how does Restore key alone affect things?

**Reply by AussieSmitty** [reply]
I hit the RESTORE key then ran the code and all still works fine.

**Reply by Whammo** [reply]
It's all good, I was going to sort this out at at some point, anyway. I'm glad to be a part of it.

**Reply by AussieSmitty** [reply]
My next challenge (to myself) is to be able to process bytes coming into DurexForth. This just requires me to read the RS232 input buffer. I'm going to assume bytes can fill the buffer in the background as it's handled by the NMI routine.

This will be fun for my planned robotic projects in Forth (DurexForth on the C64 and a version of Forth on the Arduino [and other devices should I use others in the future]).

Thanks again Whammo!

Regards,
Steve

**Reply by Whammo** [reply]
We've done a good thing.
 Maybe a half-dozen people in the world would be really excited if they knew about it. 👍

**Reply by AussieSmitty** [reply]
Hi Whammo,

The retro scene from what I'm seeing is growing, and Forth is perfect for doing robotic projects, so hopefully supporting RS232 will be more and more appreciated. I certainly am, so thanks again.

Regards,
Steve

**Reply by Whammo** [reply]
It's definitely opening a can of worms. You'll know soon how the Kernal handles nulls received, and other fun stuff.
It's also a technical challenge doing interrupt programming. Along with the opportunity to determine how the differing sources are to be handled. 
 It's especially interesting in the way the software interrupt handler ( brk ) in intertwined with the NMI handler.
 Still, up there in the thread earlier most of the work was already done.

**Comment by jkotlinski** [comment]
Sorry for not being very responsive in this thread.
I would welcome a pull request with necessary changes to add RS-232 support
in master :-)

**Reply by Whammo** [reply]
Sorry you missed out on all the fun!
 I can work something up after the next release...  ;)

**Comment by jkotlinski** [comment]
Ah, do you think there should be a release? I guess I should just package
whats on master right now. Its probably relatively OK.

**Reply by Whammo** [reply]
Are you interested in a pull request for the forth port of the DOS/io/open libraries?

**Comment by jkotlinski** [comment]
It depends. What is the size difference? I remember doing something similar
before, but found there were no savings in having it in Forth.

**Reply by Whammo** [reply]
You're correct as usual. Forth compiles to considerably more object code. It was still fun doing it.
   So as far as the rs232 routines, we also have Restore, run/stop restore and brk with which to contend.
Currently, they all fall through to QUIT. 
  The simplest solution is to have Restore fall through the NMI vector without effect, and then run/stop restore and brk goes to QUIT.
 There doesn't seem to be a good way to check for Restore other than eliminating all the other sources of NMI.

**Comment by Whammo** [comment]
Or rather it would be that Restore would fall through harmlessly when NMI is enabled, and otherwise QUIT.
  It's actually easy to eliminate all other sources of NMI, if the CIA's interrupt control register doesn't have the high bit set, ( in YR )
it's Restore or run/stop restore. That's it besides some cart or another causing NMI.

**Comment by AussieSmitty** [comment]
Hi Whammo and anyone else interested,

After confirming I could send bytes over RS232 using DurexForth, I then proceeded to check that I could also recieve bytes over RS232 also using DurexForth. After much trial and many errors, I finally have low level Forth words for sending a byte and recieving a byte (if one is present).

![DurexForth RS232](https://user-images.githubusercontent.com/41884798/124114289-c0942b80-daaf-11eb-8263-98baa813e71f.JPG)

 RS232 on C64 testing )

: setup ( -- ) 
14 53280 c! \ light blue border
154 emit    \ text colour
6 53281 c!  \ background dark blue
0 $cc c! [  \ cursor blink on
$fb stx,    \ store a copy of xreg
8 lda,#     \ low byte of params
$fd sta,    \ store at $fd
0 lda,#     \ hi byte of params
$fe sta,    \ store at $fe
$2 lda,#     \ length of filename
$fd ldx,#   \ lo byte params add
$0 ldy,#    \ hi byte params add
$ffbd jsr,  \ call kernal setnam
2 lda,#     \ 2 = RS232 device#
tax,        \ set device=2
0 ldy,#     \ no secondary #
$ffba jsr,  \ setlfs
$ffc0 jsr,  \ open
$fb ldx, ]  \ restore xreg
147 emit ;  \ clear screen

: sendbyte ( byte -- ) $fc c! [
$fb stx,    \ save xreg
2 ldx,#     \ rs232 device #
$ffc9 jsr,  \ chkout
$fc lda,	\ byte to send
$ffd2 jsr,  \ chrout
3 ldx,#     \ screen device #
$ffc9 jsr,  \ chkout
$fb ldx, ] ; \ restore xreg

: rec ( -- b f ) [ \ rs232 byte flag
$fb stx,	\ save xreg
0 lda,# 	\ clear flag byte recieved
$fd sta,	\ store at $fd
2 ldx,#		\ change to RS232
$ffc6 jsr,	\ call Kernal CHKIN
667 lda,    \ read input buffer pointer
668 cmp,    \ compare with 668 
+branch beq, \ jmp ahead if no byte
$fd inc,	\ set flag
$ffe4 jsr,	\ GETIN RS232 buffer
$fc  sta,	\ Store in $fc 
:+
0 ldx,#     \ keyboard device #
$ffc6 jsr,	\ call Kernal CHKIN
$fb ldx, ]  \ restore xreg
$fc c@ $fd c@ ;    \ byte & flag to stack 

Now to start doing some robotics via RS232 from a C64!

Regards,
Smitty
